### ISSUE TITLE
Add metrics endpoint

### ISSUE THREAD

**Issue by fitzthum** [confidential-containers/trustee#543]
Currently the only way to figure out the status of Trustee is to look at the logs. We should add a more sophisticated metrics infrastructure that can be connected into third-party systems.

**Comment by mkulke** [comment]
Note that while prometheus support is a good idea, it will not supplement logs. metrics are about aggregate values (think: number of successful tdx verifications per minute), you shouldn't/cannot debug individual transactions with it (because of restrictions on the cardinality of metrics).

the closest to a supplement logs today are opentelemetry traces. sadly, this is still a bit rough for rust webservices today. but embracing structured logs via [tracing ](https://crates.io/crates/tracing) is a step in the right direction. e.g. there should be correlation ids to debug logical transactions (like an RCAR ceremony).

**Comment by pmores** [comment]
Hi, I'm looking into adding Prometheus support.  I have a bunch of questions:

- where do we want Prometheus exporters added?  KBS?  Attestation service?  Both?  Any others?  More generally, is there any preexisting idea of the general structure of Prometheus support?
- is it acceptable to use a crate, probably [this one](https://crates.io/crates/prometheus) (it looks fairly well-maintained and widely-used), in an implementation?
- if so, is it OK to use it directly in trustee code, or does it need to be hidden behind an interface to ease re-implementing Prometheus support using a different crate if need be?  (Looking at trustee code it probably is but asking anyway to make sure.)
- could you advice a bunch of example specific metrics for each exporter to be added (preferably one on each data types or at least a counter, gauge and histogram) so that I can work with real-world use cases?

Thanks in advance!

**Comment by pmores** [comment]
Excellent, thanks Tobin!  This looks like enough for me to start an implementation.

As for a metrics abstraction, I'd actually prefer to have one too but I think I'll start by deploying the prometheus crate directly and see how it turns out.  Maybe a neat abstract interface suggests itself, or maybe I find out that I'd be just tediously replicating the underlying crate interface one-to-one.  I'm planning to post a WIP PR as early as it makes sense so we can then make a final decision on this as a community.

**Comment by spotlesstofu** [comment]
Following up @pmores foundational work, I'd like to propose which metrics to implement.

Some of these are already in the code but they're missing the suggested labels.

Metric Name | Prometheus Type | Description | Suggested Labels | Rationale/Prometheus Best Practice
-- | -- | -- | -- | --
trustee_up | Gauge | Indicates if trustee instance is healthy (1) or unhealthy (0). | None | Fundamental health check; preferred over 5xx errors.
trustee_info | Gauge | Static metadata about the trustee instance. | version, commit, build_date | Exposes non-numeric info via labels.
trustee_requests_total | Counter | Total number of logical operations/API calls processed. | endpoint, method, status_code/result | Fundamental throughput metric; uses _total suffix.
trustee_request_errors_total | Counter | Total number of failed logical operations/API calls. | endpoint, method, error_type | Complements _requests_total for error rate; separate success/failure counts.
trustee_request_duration_seconds | Histogram | Observes duration of key operations/API calls in seconds. | endpoint, method | Critical for performance monitoring; provides latency distribution.
trustee_active_connections | Gauge | Current number of active network connections/sessions. | protocol, state | Indicates current load and concurrent resource utilization.
trustee_attestation_requests_total | Counter | Total attestation requests received and processed. | attest_type, source, result | Core business logic metric for trustee.
trustee_attestation_failures_total | Counter | Total number of failed attestation requests. | attest_type, failure_reason, source | Critical for attestation reliability/security; complements _requests_total.
trustee_policy_evaluations_total | Counter | Total number of policy evaluations performed. | policy_id, result | Key metric for policy enforcement activity.
trustee_policy_violations_total | Counter | Total number of detected policy violations. | policy_id, violation_type | Direct security-related metric for non-compliance.
trustee_key_provisioning_requests_total | Counter | Total requests for cryptographic key provisioning/retrieval. | key_id, operation, status | Tracks key management activity.
trustee_key_provisioning_duration_seconds | Histogram | Measures duration of key provisioning operations. | key_id, operation | Performance of security-critical key operations.

**Comment by spotlesstofu** [comment]
Also I would like to cite the Prometheus docs with something to watch out for during the implementation:

> Be consistent in whether you count queries when they start or when they end.
> When they end is suggested, as it will line up with the error and latency stats, and tends to be easier to code.

**Comment by pmores** [comment]
With this unusually worked-out proposal (amazing job @spotlesstofu , thanks!), let's discuss design _before_ an implementation is submitted for review. @mkulke @fitzthum @Xynnn007 

Is everybody fine with the suggested labelling?

Personally, I'm not sure whether the first two (trustee_up and trustee_info) make good Prometheus metrics.  To be clear I'm not suggesting they don't - when I say "I'm not sure" here I mean it literally.

Also, probably a stupid question but anyway: are the key provisioning requests a subset of trustee requests?  If so then what's the benefit of tracking the subset separately?

**Comment by fitzthum** [comment]
> Does Trustee have internal caches or queues? 

Not yet

**Comment by fitzthum** [comment]
Regarding `trustee_key_provisioning_requests_total`, we should probably talk about `resources` rather than `keys`. Also, the term provisioning typically refers to the admin side where the resources are uploaded to Trustee. I'm not sure if this metric is referring to that or to a `get_resource` call.

For the metrics involving the policy, you should specify whether it is the resource/kbs policy or the attestation policy. 

Otherwise this seems good

**Comment by spotlesstofu** [comment]
>Regarding trustee_key_provisioning_requests_total, we should probably talk about resources rather than keys. Also, the term provisioning typically refers to the admin side where the resources are uploaded to Trustee. I'm not sure if this metric is referring to that or to a get_resource call.

The rationale for this metric is intended to observe trustee's operations as a "resource database". This involves instrumenting all create, read, update, and delete operations for resources. This includes actions from all sides, such as the admin uploading resources and the get_resource call.

A possible name could be `trustee_resource_management_` instead of `trustee_key_provisioning_`.

**Comment by pmores** [comment]
To be a bit more concrete, please check out rough implementations of the proposed resource policy & attestation metrics in PR #816.  Please take a look at how the verifier return value handling in attestation is made a bit more complex in order to capture a failure and update metric accordingly. Please also consider the semantics of especially `resource_policy_violations_total` and `attestation_failures_total` and check if it matches your expectations.

**Comment by spotlesstofu** [comment]
> what counts as a failure?

In the current scope, anything that is not success counts as a failure, no matter the reason.
Looking at the example, that includes both 2 and 3.

Increased granularity may be desired in the future, but I'd like that requirement to come from the field rather than from making assumptions.

About the latest two comments, we want that any error happening is counted as a failure. Could the failure counter be in the caller function or anywhere higher up in the call stack, closer to the API server?

**Comment by pmores** [comment]
Let me perhaps make our options (at least as I see them) explicit here to make the discussion more specific.  I believe that full reporting for attestation and policy would mean having four counters for each:
- {policy,attestation} total
- {policy,attestation} successes
- {policy,attestation} rejections
- {policy,attestation} errors

I think we can reduce this to three without losing information by assuming that `total = successes + rejections + errors` and leaving out any of the four since it can be computed from the remaining three.

We could also reduce it further to two at the cost of information loss if acceptable.  Some variants from this category are discussed above:
- keep `total`, introduce `failures = rejections + errors` => lose ability to differentiate between `rejections` and `errors` but still be able to obtain `successes` as `total - failures`, or
- keep `total`, introduce `failures = rejections` => lose ability to differentiate between `successes` and `errors` but still be able to get their sum as `total - failures`

In the context of this framework, which variant would make for the best metrics design?

**Comment by Xynnn007** [comment]
> type provides enough resolution at least to be able to tell attestation errors from policy (and other) errors

That's the key point, now we do not have such good resolution underlying. If you need, I could help to promote this -

**Comment by pmores** [comment]
@Xynnn007 I think whether we need better resolution in `Error` depends on which way we decide to go in terms of options laid out [above](https://github.com/confidential-containers/trustee/issues/543#issuecomment-2954933520).  For instance, if we want to go with @spotlesstofu's proposal where any result but successful attestation and approval by policy counts as a failure, I guess we won't need to change `Error` as that can be done right at the highest level directly in `ApiServer`, no improved error resolution required.

**Comment by Xynnn007** [comment]
@pmores Yup, I agree that rejection and error should be separated, and the indicator consumer can cluster it at its level - In addition, what I mean is that error would be different cases, we can further refine the error type by adding context and which could actually be a future work

**Comment by pmores** [comment]
@Xynnn007 cool, so if we decide to count rejection and errors separately which I'd prefer, too, and which I also believe doesn't go against @spotlesstofu's preference either since her semantics of failures can still be easily obtained as `rejections + errors`, then adding resolution to `Error` would probably make error counting easier.

For completeness, I think there's an alternative approach too, based on the `inspect_err()` call you proposed in review of #816 (e.g. we'd basically just add an `.inspect_err(|_| ATTESTATION_ERRORS.inc())` call to all '?' statements in `__attest()`) which I think would work too, though less elegantly than in `error_response()`.

> That's the key point, now we do not have such good resolution underlying. If you need, I could help to promote this -

If we decide to go with the `error_response()` solution I sure would appreciate your assistance with adapting `Error`. :-)

**Comment by pmores** [comment]
To summarise my understanding of the discussion so far:
- we want to label by TEE type by that name since even though we recognise that the name might be prone to future change there's no consensus yet on a better terminology and we don't want to preempt that consensus at this point
- we can't label by request source since that information is not available to the KBS
- there doesn't seem to be much point in labelling by policy ID since 1) policy is fixed and implicit for each KBS run, and 2) we don't have a policy ID anyway.

Please object if I got anything wrong.

In terms of dealing with Prometheus cardinality problem, considering Magnus's confirmation that the problem can actually be quite severe I'm afraid we also need additional caution labelling by anything that's supplied by KBS client.  This seems particularly relevant with respect to request source - even if we do make it available to KBS at some point, we need to consider that if for instance a malicious client floods a KBS instance with requests having randomly generated source IDs it could constitute a DoS attack on the Prometheus instance scraping the KBS instance's metrics.

(In this context I think I'll need to revisit the resource path label on our existing resource access metrics.  I'm thinking of maybe splitting those counters to successes and failures as well and only labelling successful accesses, otherwise we might be open to a similar DoS attack by a client requesting randomly generated invalid resource paths. @mkulke)

**Comment by Xynnn007** [comment]
@pmores I am not sure we have a perfect mechanism for now. Programmingly, it's now hard to check if the connection to AS is health. We'd either add a health check API to AS and let KBS call it regularly, or we can only mark it healthy once a request has been answered (not timely)

**Comment by spotlesstofu** [comment]
Reading the docs again, even if we find a way to provide a single metric of the overall health, it shouldn't be called `trustee_up`.

`_up` is already being used by Prometheus to signal when a scrape fails. See:
- https://prometheus.io/docs/instrumenting/writing_exporters/#failed-scrapes.
- https://prometheus.io/docs/concepts/jobs_instances/#automatically-generated-labels-and-time-series

In conclusion, given the two previous comments, I'd say a single liveness metric doesn't apply at the moment, however we call it.

**Comment by fitzthum** [comment]
`kbs` sounds good to me.

**Comment by pmores** [comment]
Frankly I don't know if I'm the right person to do that.  Maybe it's because I'm the implementor but everything seems rather clear and simple to me and I don't see much to document.  Metrics are served at the well-known default endpoint and they are (self-)documented the Prometheus way.

**Comment by fitzthum** [comment]
Well the average person who comes to the docs on [the website](https://confidentialcontainers.org/docs/attestation/) won't know that we even have prometheus support. It would be good to show how you can find the endpoint, spin up prometheus alongside coco, and look at the metrics.

**Comment by pmores** [comment]
I agree that saying explicitly that Prometheus endpoint is available for KBS and how to reach it makes a lot of sense though I'm not sure what the right place is for that.  For the rest of what you mention though, frankly I'm not sure myself how exactly to do that.

**Comment by fitzthum** [comment]
There is still zero documentation for the metrics endpoint, so this feature is not quite supported, and this issue remains open.
